src/org/freenetproject/freemail/MailMessage.java

```diff
@@ -576,4 +582,342 @@ public MailMessageHeader(String n, String v) {
 			this.val = v;
 		}
 	}
+
+	private static class MessageBodyReader extends BufferedReader {
+		private final Charset charset;
+		private final ContentTransferEncoding transferEncoding;
+
+		/**
+		 * Used to store data e.g. when the hard line break is in the middle of
+		 * an encoded line as can happen in e.g. base64.
+		 */
+		private String buffer;
+
+		public MessageBodyReader(Reader in, MailMessage msg) throws UnsupportedEncodingException {
+			super(in);
+			transferEncoding = ContentTransferEncoding.parse(msg.getFirstHeader("Content-Transfer-Encoding"));
+
+			String contentType = msg.getFirstHeader("Content-Type");
+			if(contentType == null) {
+				contentType = "text/plain; charset=us-ascii";
+			}
+			MediaType parsedType;
+			try {
+                parsedType = new MediaType(contentType);
+            } catch (MalformedURLException e) {
+                throw new UnsupportedEncodingException("Can't handle content type: \""+contentType+"\" : "+e.getMessage());
+            }
+            if(!(parsedType.getType().equals("text") && parsedType.getSubtype().equals("plain"))) {
+                throw new UnsupportedEncodingException("Can't handle content types other than text/plain. Type was "
+                        + parsedType);
+            }
+            String charsetName = parsedType.getParameter("charset");
+            if(charsetName == null) charsetName = "us-ascii";
+            Map<String,String> params = parsedType.getParameters();
+            params.remove("charset");
+            if(!params.isEmpty()) {
+				throw new UnsupportedEncodingException("Can't handle text/plain with parameter other than charset. "
+						+ "Unrecognised parameters were: " + params);
+			}
+
+			charset = Charset.forName(charsetName);
+		}
+
+		@Override
+		public String readLine() throws IOException {
+			switch(transferEncoding) {
+			case BASE64:
+				return readBase64Line();
+			case QUOTED_PRINTABLE:
+				return readQpLine();
+			case SEVEN_BIT:
+				return super.readLine();
+			default:
+				Logger.error(this, "Missing case in transfer encoding switch: " + transferEncoding);
+				assert (false);
+				return super.readLine();
+			}
+		}
+
+		/**
+		 * Reads and decodes quoted-printable data until a canonical line has
+		 * been decoded.
+		 * @return a canonical line of message text
+		 * @throws IOException if an I/O error occurs while reading the input
+		 */
+		private String readQpLine() throws IOException {
```

Comment:
These methods should totally be broken out of here into their own classes, with their own tests and all.